source/slang/slang-ir-metal-legalize.cpp

```diff
@@ -731,67 +1393,105 @@ namespace Slang
-                    auto groupIndexCalc = emitCalcGroupThreadIndex(svBuilder, groupThreadId, computeExtent);
+                    auto groupIndexCalc = emitCalcGroupIndex(svBuilder, entryPointToGroupThreadId[entryPoint.entryPointFunc], computeExtent);
                     svBuilder.addNameHintDecoration(groupIndexCalc, UnownedStringSlice("sv_groupindex"));
 
-                    param->replaceUsesWith(groupIndexCalc);
-                    param->removeAndDeallocate();
+                    var->replaceUsesWith(groupIndexCalc);
+                    var->removeAndDeallocate();
                 }
             }
             if (info.isUnsupported)
             {
-                reportUnsupportedSystemAttribute(sink, param, semanticName);
-                continue;
+                reportUnsupportedSystemAttribute(var, semanticName);
+                return;
             }
-            if (!info.requiredType)
-                continue;
+            if (!info.permittedTypes.getCount())
+                return;
 
-            builder.addTargetSystemValueDecoration(param, info.metalSystemValueName.getUnownedSlice());
+            builder.addTargetSystemValueDecoration(var, info.metalSystemValueName.getUnownedSlice());
 
-            // If the required type is different from the actual type, we need to insert a conversion.
-            if (info.requiredType != param->getFullType() && info.altRequiredType != param->getFullType())
+            bool varTypeIsPermitted = false;
+            auto varType = var->getFullType();
+            for (auto& permittedType : info.permittedTypes)
             {
-                auto targetType = param->getFullType();
-                builder.setInsertBefore(entryPoint.entryPointFunc->getFirstBlock()->getFirstOrdinaryInst());
-                param->setFullType(info.requiredType);
-                List<IRUse*> uses;
-                for (auto use = param->firstUse; use; use = use->nextUse)
-                    uses.add(use);
-                auto convertedValue = tryConvertValue(builder, param, targetType);
-                copyNameHintAndDebugDecorations(convertedValue, param);
-                if (!convertedValue)
-                {
-                    // If we can't convert the value, report an error.
-                    StringBuilder typeNameSB;
-                    getTypeNameHint(typeNameSB, info.requiredType);
-                    sink->diagnose(param->sourceLoc, Diagnostics::systemValueTypeIncompatible, semanticName, typeNameSB.produceString());
-                }
-                else
+                varTypeIsPermitted = varTypeIsPermitted || permittedType == varType;
+            }
+
+            if (!varTypeIsPermitted)
+            {
+                // Note: we do not currently prefer vector<T,N> to vector<P,N>.
+                // vector<T,N> to vector<T,N-1> is equally preferred.
+                bool foundAConversion = false;
```

Comment:
why can't we convert uint3 to int3?

Reply:
We can convert uint3 to int3

The following comment was more about addressing the following example:
system value example allowed types: `float4`, `uint4`, `int4`
user used: `float2`
Slang will equally prefer `float4` to `uint4` to `int4`. This means the type may lose data if slang selects `uint4` or `int4`.

I will update the comment for clarity